src/coreclr/vm/debugdebugger.cpp

```diff
@@ -1013,14 +1013,180 @@ void DebugStackTrace::Element::InitPass1(
 }
 
 #ifndef DACCESS_COMPILE
+// This is an implementation of a cache of the Native->IL offset mappings used by managed stack traces. It exists for the following reasons:
+// 1. When a large server experiences a large number of exceptions due to some other system failing, it can cause a tremendous number of stack traces to be generated, if customers are attempting to log.
+// 2. The native->IL offset mapping is somewhat expensive to compute, and it is not necessary to compute it repeatedly for the same IP.
+// 3. Often when these mappings are needed, the system is under stress, and throwing on MANY different threads with similar callstacks, so the cost of having locking around the cache may be significant.
+//
+// The cache is implemented as a simple hash table, where the key is the IP + fAdjustOffset
+// flag, and the value is the IL offset. We use a version number to indicate when the cache
+// is being updated, and to indicate that a found value is valid, and we use a simple linear
+// probing algorithm to find the entry in the cache.
+//
+// The replacement policy is randomized, and there are s_stackWalkCacheWalk(8) possible buckets to check before giving up.
+//
+// Since the cache entries are greater than a single pointer, we use a simple version locking scheme to protect readers.
+
+struct StackWalkNativeToILCacheEntry
+{
+    void* ip = NULL; // The IP of the native code
+    uint32_t ilOffset = 0; // The IL offset, with the adjust offset flag set if the native offset was adjusted by STACKWALK_CONTROLPC_ADJUST_OFFSET
+};
+
+static LONG s_stackWalkNativeToILCacheVersion = 0;
+static DWORD s_stackWalkCacheSize = 0; // This is the total size of the cache (We use a pointer+4 bytes for each entry, so on 64bit platforms 12KB of memory)
+const DWORD s_stackWalkCacheWalk = 8; // Walk up to this many entries in the cache before giving up
+const DWORD s_stackWalkCacheAdjustOffsetFlag = 0x80000000; // 2^31, put into the IL offset portion of the cache entry to check if the native offset was adjusted by STACKWALK_CONTROLPC_ADJUST_OFFSET
+static StackWalkNativeToILCacheEntry* s_stackWalkCache = NULL;
+
+bool CheckNativeToILCacheCore(void* ip, bool fAdjustOffset, uint32_t* pILOffset)
+{
+    // Check the cache for the IP
+    int hashCode = MixPointerIntoHash(ip);
+    StackWalkNativeToILCacheEntry* cacheTable = VolatileLoad(&s_stackWalkCache);
+    
+    if (cacheTable == NULL)
+    {
+        // Cache is not initialized
+        return false;
+    }
+    DWORD cacheSize = VolatileLoadWithoutBarrier(&s_stackWalkCacheSize);
```

Comment:
If we tested cacheSize != 0 it looks like we could eliminate all the VolatileStore/Load for s_stackWalkCache and s_stackWalkCacheSize. Regardless of read/write ordering the variables only have two possible values, 0 or initialized. I am mostly just trying to minimize the amount of memory ordering that someone has to reason about when trying to understand the multi-threaded behavior of this cache.